torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1079,13 +1095,96 @@ def _check_wrapped(cls, begin_module, check_fn, err_fn):
             if not check_fn(mod):
                 raise ValueError(err_fn(mod))
 
+    # TODO (awgu): I do not think we need this -- it is only needed if
+    # `_apply()` is not in-place (which is the subroutine for `to()`, `cuda()`,
+    # etc.)
+    # def _apply(self, *args, **kwargs):
+    #     fsdp = super()._apply(*args, **kwargs)
+    #     for param in fsdp.params:
+    #         handle = param._handle
+    #         assert handle in fsdp._handles
+    #         handle.flat_param = param
+    #     return fsdp
+
+
+    def _register_param_handle(self, handle: FlatParamHandle) -> None:
+        """Registers the parameter handle to this FSDP instance."""
+        if handle not in self._handles:
+            self._handles.append(handle)
+
+    def _register_param_handle_from_params(
+        self,
+        params: List[nn.Parameter],
+        root_module: nn.Module,
+    ) -> None:
+        """Constructs a parameter handle with the given ``params`` and
+        registers it to this FSDP instance."""
+        if len(params) == 0:
+            raise ValueError("`params` must be non-empty")
+        handle = FlatParamHandle(params, root_module)
+        for module in handle._get_modules():
+            self._module_to_handles[module].append(handle)
+        self._register_param_handle(handle)
+
+    def _register_param_handles_from_module(
+        self,
+        root_module: nn.Module,
+        bucket_size: int,
+    ) -> None:
+        """
+        This should only be called once. For now, this is called in the FSDP
+        constructor but should be changed to be in the post-backward callback
+        of the first iteration.
+
+        Defines:
+            _fsdp_wrapped_module
+            _module_to_handles: A mapping from original module to the handles
+                that contain some part of its parameters.
+            params: A list of all :class:`FlatParameter` s.
+            _module_to_forward: A mapping from original module to the module's
+                unpatched ``forward()`` method.
+        """
+        # For the non-recursive wrapping, we set TODO
+        self._fsdp_wrapped_module = root_module
+        self._module_to_handles: Dict[nn.Module, List[FlatParamHandle]] = \
+            collections.defaultdict(list)
+
+        # TODO (awgu): For now, we assume a fixed bucket size and follow
```

Comment:
@LinjianMa  I am doing the handle creation in the FSDP constructor right now. We will need to move this to after the first iteration (probably in the post-backward callback).

This just goes in `.parameters()` order and buckets according to some threshold. This is good for testing since I can make the bucket threshold small to force many handles to be created.